public/app/features/variables/state/keyedVariablesReducer.ts

```diff
@@ -0,0 +1,50 @@
+import { AnyAction } from 'redux';
+import { createAction, PayloadAction } from '@reduxjs/toolkit';
+import { getTemplatingReducers, TemplatingState } from './reducers';
+import { variablesInitTransaction } from './transactionReducer';
+import { toStateKey } from '../utils';
+
+export interface KeyedVariablesState {
+  lastKey?: string;
+  keys: Record<string, TemplatingState>;
+}
+
+export const initialKeyedVariablesState: KeyedVariablesState = { keys: {} };
+
+export interface KeyedAction {
+  key: string;
+  action: PayloadAction<any>;
+}
+
+const keyedAction = createAction<KeyedAction>('templating/keyedAction');
+
+export function toKeyedAction(key: string, action: PayloadAction<any>): PayloadAction<KeyedAction> {
```

Comment:
nitpick / future fear. Since this name is so generic, we might create a new function like this (to use this pattern for other stores like dashboard), and the signature expects any payload it could be easy to get the wrong toKeyedAction. 

Alternative
* toVariablesSystemAction 
* a bit long... any others?

Reply:
That's true but the reducers wrapped by this will still only act on actions they know right? But from a performance point of view I can see how naming it differently is a good idea.

Reply:
I might have an idea to use a union type instead of `any` to solve this. Tracking this in another issue #44769.